Describe pinned env (#406)
aeternity/protocol: node/api/channels_api_usage.md

Only steps 1 and 4 require chain interactions, step 2 and 3 are off-chain.

> ### On-chain requests requests

Reviewer: ```suggestion
### On-chain requests
```

There are two types of requests:
* Total amount-modifying ones - [deposit](#deposit-transaction) and [withdrawal](#withdraw-transaction)
* [Channel-closing ones](#solo-closing-sequence) - [solo close](#solo-close-on-chain-transaction), [slash](#slash-on-chain-transaction) and [settle](#settle-on-chain-transaction)

### Pinned environment

While on-chain consensus is reached between miners, in off-chain world we
don't have those. State channels are two-party systems that are closer to
proof of stake solutions with the remark that both participants have equal
stake in the channel, no matter their balances. The channel can make another
step forward only if both parties agree upon the new state or it is forced
progressed on-chain based upon a previous mutually agreed state. This makes
channels both trustless and egalitarian.

This trustless model is based upon both participants executing off-chain
updates locally and reaching the same results. This is how consensus is

> reached betweenn them. Since off-chain smart contracts can read on-chain

Reviewer: ```suggestion
reached between them. Since off-chain smart contracts can read on-chain
```

objects like accounts, names, contracts and oracles requests and responses,
the results of their execution rely heavily on the chain they had been based
on.

Participants are expected to use their own nodes to support their channels.

> At the moment using a service hosted by a third party is trustfull and thus

Reviewer: ```suggestion
At the moment using a service hosted by a third party is trustful, thus
```

> discouraged. This leads to both participants' nodes being peers in a system

Reviewer 2: ```suggestion
discouraged. This leads to both participant's nodes being peers in a system
```

with an eventual consistency - due to network constraints and forks both
participants can have a different view of the chain.

The combination of participants having different views on the chain and the
off-chain consensus being dependent on it could lead to a fragile system with
a lot of mismatching state hashes of off-chain updates. In order to improve
this there is an optional functionality of setting `block_hash` that defines
the on-chain environment that the update is to be executed at. We call this
shared view of the chain _a pinnned environment_. When a participant wants to
start a new round of updates, one can optionally specify a pinned environment

> to be ran at. This is how it communicates to the other party what one

Reviewer 2: ```suggestion
to be ran at. This is how the participant communicates to the other party what one
```

considers to be a block hash that is safe enough to base an off-chain update
upon. The other party might decide if the block hash is too old or too new
depending on their local view of the chain. If the specified pinned

> environment does not meet the expectations, the whole update is reject as an

Reviewer: ```suggestion
environment does not meet the expectations, the whole update is rejected as
```

> invalid one.

Reviewer: ```suggestion
invalid.
```

An update might not be pinned to any environment. In that case a placeholder
value for the blockhash is provided:
`"kh_11111111111111111111111111111111273Yts"` or
`"mh_11111111111111111111111111111111273Yts"`. In this case both participants

> use the whatever they see to be the latest top block.

Reviewer: ```suggestion
use whatever they see to be the latest top block.
```

> The `block_hash` is an optional argument to all mutual transactions. If it is

Reviewer 2: ```suggestion
The `block_hash` is an optional argument to all mutual offchain transactions. If it is
```

  We will be using the tool [wscat](https://github.com/websockets/wscat)
  We assume the channel's WebSocket listener is set on port 3014 (default one)

#### Channel block hash delta values

>   A client can specify what are considered by them to be a valid block hash.

Reviewer: ```suggestion
  A client can specify what is considered by them to be a valid block hash.
```

  Those are defined as deltas according to the latest chain top as seen from

>   the participant's node. Delta of 0 is the latest top, delta of 1 is the

Reviewer: ```suggestion
  the participant's node. A delta of `0` is the latest top, a delta of `1` is the
```

>   previous generation and etc. Setting the deltas would result in block hashes

Reviewer: ```suggestion
  previous generation, etc. Setting the deltas would result in block hashes
```

>   provided that are newer than the specified value for maturity to be

Reviewer: This sentence doesn't make sense.
